### PR TITLE
Child teiid core

### PR DESCRIPTION
**Pull request by aditya300899**
@rareddy kindly review. their was error in the ManagedConnectionFactoryClass but that would be sorted once the class resource connection by shawkins is used at place of basic connection.

### CODE REVIEW

**Comment by rareddy** [comment]
@shawkins can you review this? @aditya300899 is hoping to be GSOC student for Teiid over this summer.

**Comment by rareddy** [comment]
@aditya300899 can you fix the build error, looks like the dependency is missing. You can build wildfly side by executing 
```
mvn -P release,wildfly install
```

**Comment by aditya300899** [comment]
@rareddy I'd already built wildfly. Since I've worked considering the ResourceConnection class as done by shawkins in salesforce. It was throwing some error in CassandraManagedConnectionFactory class. Could it be the reason? If this is so I can do the changes as done by shawkins and change the contracts in accordance with ResourceConnection class.

File: wildfly/connectors/cassandra/connector-cassandra/pom.xml [hunk 1]
```diff
@@ -34,6 +34,12 @@
             <groupId>org.teiid.connectors</groupId>
             <artifactId>translator-cassandra</artifactId>
         </dependency>
-	</dependencies>
+        <dependency>
+            <groupId>org.teiid.connectors</groupId>
+            <artifactId>cassandra-api</artifactId>
+            <version>14.0.0-SNAPSHOT</version>
```
Comment (rareddy) [inline wildfly/connectors/cassandra/connector-cassandra/pom.xml:40]: Remove the version here, we should not directly reference the version like that

File: connectors/cassandra/cassandra-api/pom.xml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
```
Comment (rareddy) [inline connectors/cassandra/cassandra-api/pom.xml:1]: You added the api project but this is not added in the parent pom.xml file, thus the build is failing to find the `cassandra-api` module. Add this to https://github.com/teiid/teiid/blob/master/connectors/pom.xml#L91

Comment (rareddy) [reply connectors/cassandra/cassandra-api/pom.xml:1]: To package this into WildFly you need also reference here https://github.com/teiid/teiid/blob/master/wildfly/connectors/cassandra/feature-pack-cassandra/src/main/resources/modules/system/layers/dv/org/jboss/teiid/resource-adapter/cassandra/main/module.xml#L18  

This module stuff is little complicated, just add the API module there, that should be enough

File: connectors/cassandra/cassandra-api/src/main/java/org/teiid/cassandra/BaseCassandraConnection.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
+package org.teiid.cassandra;
```
Comment (rareddy) [inline connectors/cassandra/cassandra-api/src/main/java/org/teiid/cassandra/BaseCassandraConnection.java:1]: Add the license header

File: connectors/cassandra/cassandra-api/src/main/java/org/teiid/cassandra/CassandraConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package org.teiid.cassandra;
```
Comment (rareddy) [inline connectors/cassandra/cassandra-api/src/main/java/org/teiid/cassandra/CassandraConfiguration.java:1]: License header here also, every new file needs to have a license header

File: wildfly/connectors/cassandra/connector-cassandra/src/main/java/org/teiid/resource/adapter/cassandra/CassandraConnectionImpl.java [hunk 5]
```diff
@@ -18,117 +18,59 @@
...
         this.config = config;
-        this.metadata = metadata;
     }
 
     public CassandraConnectionImpl(CassandraManagedConnectionFactory config) {
+        super(getCluster(config), config.getKeyspace());
         this.config = config;
+    }
 
+    private static Cluster getCluster(CassandraManagedConnectionFactory config) {
```
Comment (rareddy) [inline wildfly/connectors/cassandra/connector-cassandra/src/main/java/org/teiid/resource/adapter/cassandra/CassandraConnectionImpl.java:50]: Looks like this method can be moved to the BaseCassandraConnection, if you pass config as "CassandraConfiguration"

File: wildfly/connectors/cassandra/connector-cassandra/src/main/java/org/teiid/resource/adapter/cassandra/CassandraConnectionImpl.java [hunk 6]
```diff
@@ -18,117 +18,59 @@
...
         Cluster.Builder builder  = Cluster.builder().addContactPoint(config.getAddress());
 
-        if (this.config.getUsername() != null) {
-            builder.withCredentials(this.config.getUsername(), this.config.getPassword());
+        if (config.getUsername() != null) {
```
Comment (rareddy) [inline wildfly/connectors/cassandra/connector-cassandra/src/main/java/org/teiid/resource/adapter/cassandra/CassandraConnectionImpl.java:53]: Looks like this whole class's methods can be moved to base class, then you just need to class definition. Make sure if you move log messages the message text also recreated in the API module.

**Comment by aditya300899** [comment]
I'll do the changes and see that i learn the things you mentioned. Thanks @rareddy

File: connectors/cassandra/cassandra-api/pom.xml (hunk 2 above)
Comment (aditya300899) [reply connectors/cassandra/cassandra-api/pom.xml:1]: you mean adding "casandra-api" to the cassandra module's pom, right? I think you have linked the wrong pom here.

File: connectors/cassandra/cassandra-api/src/main/java/org/teiid/cassandra/BaseCassandraConnection.java (hunk 3 above)
Comment (aditya300899) [reply connectors/cassandra/cassandra-api/src/main/java/org/teiid/cassandra/BaseCassandraConnection.java:1]: done

File: connectors/cassandra/cassandra-api/src/main/java/org/teiid/cassandra/CassandraConfiguration.java (hunk 4 above)
Comment (aditya300899) [reply connectors/cassandra/cassandra-api/src/main/java/org/teiid/cassandra/CassandraConfiguration.java:1]: got it, done

File: wildfly/connectors/cassandra/connector-cassandra/pom.xml (hunk 1 above)
Comment (aditya300899) [reply wildfly/connectors/cassandra/connector-cassandra/pom.xml:40]: when i remove version, the maven declares the dependency to be unknown. I tried a few things to work this out but couldn't do. How to do this?

File: wildfly/connectors/cassandra/connector-cassandra/src/main/java/org/teiid/resource/adapter/cassandra/CassandraConnectionImpl.java (hunk 6 above)
Comment (aditya300899) [reply wildfly/connectors/cassandra/connector-cassandra/src/main/java/org/teiid/resource/adapter/cassandra/CassandraConnectionImpl.java:53]: I've done this. Just be sure, you meant referencing CassandraManagedConnectionFactory's object with the CassandraConfiguration interface, right?
Also I couldn't understand the log messages thing. What I understand is that you mean "the same message text need to be there when doing it in the api module in BaseCassandraConnection " but i don't know how to make the managedconnectionfactory class field "UTIL" to be made available. Also, is it necessary?

File: wildfly/connectors/cassandra/connector-cassandra/pom.xml (hunk 1 above)
Comment (rareddy) [reply wildfly/connectors/cassandra/connector-cassandra/pom.xml:40]: It needs to go into parent pom under <DepencyManagement> section.

File: wildfly/connectors/cassandra/connector-cassandra/src/main/java/org/teiid/resource/adapter/cassandra/CassandraConnectionImpl.java (hunk 5 above)
Comment (rareddy) [reply wildfly/connectors/cassandra/connector-cassandra/src/main/java/org/teiid/resource/adapter/cassandra/CassandraConnectionImpl.java:50]: I am saying the getCluster method can be moved into base class

File: wildfly/connectors/cassandra/connector-cassandra/pom.xml (hunk 1 above)
Comment (rareddy) [reply wildfly/connectors/cassandra/connector-cassandra/pom.xml:40]: May be this one I think https://github.com/teiid/teiid/blob/master/connectors/cassandra/pom.xml

File: connectors/cassandra/cassandra-api/pom.xml (hunk 2 above)
Comment (rareddy) [reply connectors/cassandra/cassandra-api/pom.xml:1]: yes

File: wildfly/connectors/cassandra/connector-cassandra/src/main/java/org/teiid/resource/adapter/cassandra/CassandraConnectionImpl.java (hunk 6 above)
Comment (rareddy) [reply wildfly/connectors/cassandra/connector-cassandra/src/main/java/org/teiid/resource/adapter/cassandra/CassandraConnectionImpl.java:53]: if you moved any log messages to another module, then externalized strings in logs there do not have property file that goes with it. see in the "resources" folder there is i18n.properties file. I am saying part of those that are affected also needs to be moved.

File: wildfly/connectors/cassandra/connector-cassandra/pom.xml (hunk 1 above)
Comment (aditya300899) [reply wildfly/connectors/cassandra/connector-cassandra/pom.xml:40]: i think you are telling this for the above comment about addition of cassandra-api module.
What is to be done about the version thing?

Comment (rareddy) [reply wildfly/connectors/cassandra/connector-cassandra/pom.xml:40]: You need to add with the version here https://github.com/teiid/teiid/blob/master/pom.xml#L1635 then leave above as is without version.

**Comment by aditya300899** [comment]
@rareddy why is the build failing now. In my opinion I've made all the changes that you had suggested.

File: connectors/cassandra/cassandra-api/pom.xml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+            <artifactId>teiid-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.teiid</groupId>
+            <artifactId>teiid-common-core</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.jboss.spec.javax.resource</groupId>
+            <artifactId>jboss-connector-api_1.7_spec</artifactId>
+        </dependency>
```
Comment (rareddy) [inline connectors/cassandra/cassandra-api/pom.xml:23]: See if you can remove this dependency, as this is JEE specific jar and we do not to want bring that dependency. See what kind of errors you run into and fix

**Comment by rareddy** [comment]
Also when you click the above link it will tell you where the build is failing.

**Comment by rareddy** [comment]
The build is still failing what does it say on your IDE?

**Comment by aditya300899** [comment]
it is building fine on my IDE.

**Comment by rareddy** [comment]
did you run the `mvn clean install` on your machine? looks like some dependency issue is occurring

**Comment by aditya300899** [comment]
when i run "mvn clean install -P dev -s settings.xml" and "mvn clean install -P dev -s settings.xml", the build is successful. But, when I do mvn clean install, some of the dependencies aren't resolved.

**Comment by rareddy** [comment]
`"mvn clean install -P dev -s settings.xml"` is correct form, but same is not passing the CI build.

**Comment by aditya300899** [comment]
I'm trying to figure out what's wrong.

**Comment by aditya300899** [comment]
@rareddy it's fine now. I didn't know that having extra dependencies could also fail a build. Kindly check

**Comment by aditya300899** [comment]
@rareddy now the build is passing partially, failing at some integration tests. How to fix this?

**Comment by aditya300899** [comment]
Failed to define class org.teiid.resource.adapter.cassandra.CassandraManagedConnectionFactory in Module "org.jboss.teiid.resource-adapter.cassandra"
I think this is the major reason

**Comment by rareddy** [comment]
we need these changes https://github.com/rareddy/teiid/commit/803aeff7216b2b91443cacbcf5ea2c79717d6e90

I wanted to send you a PR, but what is your branch in git? Your commit history seems very weaved with other commits in between I could not figure out

**Comment by aditya300899** [comment]
child-teiid-core is the branch name.

**Comment by aditya300899** [comment]
1) I think I've streamlined my commit history now for future. It's not weaved anymore. Please check!
2) You can give a pr for branch child-teiid-core
3) Why is the build passing now?
4)I can see that you've done some changes in the commit that you've mentioned. What was the purpose of changing the package for translator-cassandra and what is it that we are doing in feature-pack?

**Comment by rareddy** [comment]
> I think I've streamlined my commit history now for future. It's not weaved anymore. Please check!
> You can give a pr for branch child-teiid-core

Did you apply the changes I gave, I do not see them. Pull them in. 

> Why is the build passing now?

No idea yet

>I can see that you've done some changes in the commit that you've mentioned. What was the >purpose of changing the package for translator-cassandra and what is it that we are doing in >feature-pack?

Because it was done wrong previously because there was no API module, the API should not depend on a translator, the translator should depend on API. so, I had to move the offending connection class to the API. Feature packs build the modules for the WildFly, thus integration tests were failing

**Comment by aditya300899** [comment]
The changes are now pulled in. Please check!
Also, do tell me the next task in hand.

**Comment by rareddy** [comment]
@shawkins can you check this one out and pull into the Teiid?

@aditya300899 Thank you. The next step is to write a connection factory module in the Teiid Spring Boot repository. Basically, take look any of mongodb, infinispan like components under "data" directory there and see how they are using the API and building separate connection classes that work in the Spring Boot environment. There will be a "example" project too along with it there

**Comment by aditya300899** [comment]
Thanks @rareddy for mentoring and making me learn a lot of things. I'll now begin with the next task.

**Comment by rareddy** [comment]
@aditya300899 I want to assign this to you https://issues.redhat.com/browse/TEIIDSB-57 But I did not find your userid in the JIRA. Did you create an account there on that system, if not create one and send me the user handle?

> for mentoring and making me learn a lot of things

If are inclined to learn and put hard work, I promise to make a great start to your career.

**Comment by aditya300899** [comment]
Thank you. I'll try my best to come up to your expectations.
I have an account with the handle invinci_1

**Comment by rareddy** [comment]
I can't find it. can you post a link to your profile? Also, create an account where others can easily find it. In Open Source you want to show clear names so that others can recognize you.

**Comment by aditya300899** [comment]
I understand! Actually, I had almost all my accounts with that username. Nevermind. 
I've made a new account
The details are as following:
Username:	adityamanglam
Full name:	Aditya Manglam Sharma
Email:	adityamanglamsharma1@gmail.com
The link to the profile is:
https://issues.redhat.com/secure/ViewProfile.jspa?name=adityamanglam

**Comment by rareddy** [comment]
assigned to you https://issues.redhat.com/browse/TEIIDSB-57

**Comment by shawkins** [comment]
> @shawkins can you check this one out and pull into the Teiid?

Don't hate me, but I'd vote for a structural change.  In the cases where the translator already has direct dependencies on the third-party jars and it's unlikely that we'll ever have multiple implementations, it's easier to just combine the api and translator into a single dependency.

In this case the translator already is using classes directly from  com.datastax.driver , guava, etc. So the api isn't providing a complete abstraction.  That was the approach for example with salesforce.

Does that make sense?

**Comment by rareddy** [comment]
@aditya300899 What is @shawkins is asking to move your API module packed directly in to cassarda translator and remove the entries in the feature-pack and any depencies on API module in pom file and re-submit. Can you please make those changes?

**Comment by aditya300899** [comment]
> @aditya300899 What is @shawkins is asking to move your API module packed directly in to cassarda translator and remove the entries in the feature-pack and any depencies on API module in pom file and re-submit. Can you please make those changes?

Ok, I'll probably raise a new pr then!

**Pull request closed without merging**